src/Razor/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.MvcApplicationPartsDiscovery.targets

```diff
@@ -0,0 +1,103 @@
+      <_ReferencesMicrosoftAspNetCoreApp>@(FrameworkReference->AnyHasMetadataValue('Identity', 'Microsoft.AspNetCore.App'))</_ReferencesMicrosoftAspNetCoreApp>
+    </PropertyGroup>
+  </Target>
+
+  <Target
+    Name="_DiscoverMvcApplicationParts"
+    Inputs="$(MSBuildAllProjects);$(ResolveAssemblyReferencesStateFile)"
+    Outputs="$(_MvcApplicationPartAttributeGeneratedFile)"
+    DependsOnTargets="$(_DiscoverMvcApplicationPartsDependsOn)"
+    Condition="'$(_ReferencesMicrosoftAspNetCoreApp)' == 'true'">
+
+    <ItemGroup>
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Abstractions" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.ApiExplorer" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Core" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Cors" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.DataAnnotations" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Formatters.Json" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Formatters.Xml" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Localization" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.Razor" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.RazorPages" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.TagHelpers" />
+      <_MvcAssemblyName Include="Microsoft.AspNetCore.Mvc.ViewFeatures" />
+    </ItemGroup>
+
+    <DiscoverMvcApplicationParts ResolvedReferences="@(ReferencePath)" MvcAssemblyNames="@(_MvcAssemblyName)">
+      <Output TaskParameter="ApplicationPartAssemblyNames" ItemName="_ApplicationPartAssemblyNames"/>
+    </DiscoverMvcApplicationParts>
+
+    <ItemGroup>
+       <_MvcApplicationPartAttribute Include="Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPart">
+        <_Parameter1>%(_ApplicationPartAssemblyNames.Identity)</_Parameter1>
+      </_MvcApplicationPartAttribute>
+    </ItemGroup>
+
+    <!-- If we did found application part assemblies, generate attributes for it and add it to compilation list -->
+    <WriteCodeFragment
+      AssemblyAttributes="@(_MvcApplicationPartAttribute)"
+      Language="$(Language)"
+      OutputFile="$(_MvcApplicationPartAttributeGeneratedFile)"
+      Condition="'@(_ApplicationPartAssemblyNames->Count())' != '0'" />
+
+    <ItemGroup Condition="'@(_ApplicationPartAssemblyNames->Count())' != '0'">
+      <Compile Remove="$(_MvcApplicationPartAttributeGeneratedFile)" Condition="'$(Language)'!='F#'" />
+      <Compile Include="$(_MvcApplicationPartAttributeGeneratedFile)"  Condition="'$(Language)'!='F#'" />
+
+      <CompileBefore Remove="$(_MvcApplicationPartAttributeGeneratedFile)" Condition="'$(Language)'=='F#'" />
+      <CompileBefore Include="$(_MvcApplicationPartAttributeGeneratedFile)"  Condition="'$(Language)'=='F#'" />
```

Comment:
For my own knowledge, what's `CompileBefore`? F# thing?

Reply:
Yeah. If you add it to `Compile` it complains about the entry point not being the first assembly. I based this on what the F# targets do for the SDK attributes:

https://github.com/dotnet/fsharp/blob/master/src/fsharp/FSharp.Build/Microsoft.FSharp.Overrides.NetSdk.targets#L30-L33